[FEAT] Numscript Builder (#12)
PagoPlus/numscriptex: guides/builder.md

# Builder
This is a feature that makes possible to build numscripts dynamically within your application.

## Usage
You just need to call `Numscriptex.Builder.build/1` with the metadata necessary to build your numscript.

### Metadata

> Metadata is the sole argument you need to use the `build/1` function, wich have the following fields:

Reviewer: ```suggestion
Metadata is the sole argument you need to use the `build/1` function, which has the following fields:
```

Required fields:
- splits: a list of metadata needed to build numscripts (explained below)

Optional fields:
- percent_asset: monetary asset (BRL, USD, EUR, and etc.) to use when build numscripts using percentage values.

> - remaining_to¹: where the rest of the money will go after the transaction

Reviewer: ```suggestion
- remaining_to¹: where the rest of the money will go after the transaction is made
```

> - remaining_to¹: where the rest of the money will go after the transaction

Author: não tem, é que e refiz um monte de vezes e acabei esquecendo de tirar

>   - Only needed if you want to designate an specific destination, the default is `remaining kept` wich makes the remainder go back to the source.

Reviewer: ```suggestion
  - Only needed if you want to designate a specific destination. The default is `remaining kept` wich makes the remainder go back to the source.
```

>   - Only needed if you want to designate an specific destination, the default is `remaining kept` wich makes the remainder go back to the source.

Reviewer: adicionar link para os docs de `remaining kept`

Author: infelizmente não tem, mas vou colocar isso [aqui](https://docs.formance.com/numscript/reference/destinations#allocation-destinations) que é o maximo que abordam sobre

About the "list of metadata" mentioned on the `split` field above:

> Required fields:

Reviewer: A linha não quebra. Adicionar um enter.

Reviewer: ![image](https://github.com/user-attachments/assets/9b92b5df-2c01-4658-bed9-a68076483bd3)

- account: account whose the money will go to

> - amount: amount of money transferred (accepts both integer and float)

Reviewer: explicar se é enviado como integer, se é tratado como o numero que é ou como decimal, ou como que.

- type: the amount type, accepts two values:
  - fixed: treats the `amount` field as a number
  - percent: treats the `amount` field as a percentage value

Optional fields:
Also accepts `aplits` and `remaining_to` fields, and a `asset` field:

- splits: in this case, this field is only neccessary if you want to build [nested destinations](https://docs.formance.com/numscript/reference/destinations#nested-destinations), and therefore it is only valid if the `type` field value is `:percent`
- asset: also the monetary asset, but in this case is only needed on `:fixed` types

The `remaining_to` field continues with the same behaviour as above.

### Example
With the metadata type above in mind, let's build a simple numscript.
```elixir
# metadata:
%{
  splits: [
    %{
      type: :percent,
      amount: 20,
      account: "some:destination"
    }
  ]
  remaining_to: "another:destination",

>   percent_asset: "USD/2"

Reviewer: Na documentação indica usar a moeda, mas aqui está usando moeda e quantidade de decimais.

}
```
Wich will generate the following numscript:
```
send [USD/2 *] (
  source = @user
  destination = {
    20% to @some:destination
    remaining to @another:destination
  }
)
```
If the `:remaining_to` field were not defined, the remaining would be kept:
```
send [USD/2 *] (
  source = @user
  destination = {
    20% to @some:destination
    remaining kept
  }
)
```

P.S.: The percent amount types will always be executed first, so we can assure that

> all of the percentage is based off of the initial value. This feature is still in

Reviewer: ```suggestion
all of the percentages are based off of the initial value. This feature is still in
```

development as we speak, but soon we will have more flexibility in those regards.

## Examples
### Example 1
Both fixed and percent types:
```elixir
%{
  splits: [
    %{
      type: :fixed,
      amount: 5,
      asset: "USD/2",
      account: "some:destination:a"
    },
    %{
      type: :percent,
      amount: 20,
      account: "some:destination:b"
    }
  ],
  remaining_to: "some:destination:c",
  percent_asset: "USD/2"
}
```

> Will build:

Reviewer: ```suggestion
Will generate:
```

```
send [USD/2 *] (
  source = @user
  destination = {
    20% to @some:destination:b
    remaining kept
  }
)

send [USD/2 500] (
  source = @user
  destination = @some:destination:a
)
```

### Example 2

> Nested percent type plus fixed

Reviewer: Adicionar comas

Author: virgula mesmo ou dois pontos (:)? nao entendi

Reviewer: Ideas:
- Nested percent-type plus fixed-type
- Nested percent type, plus fixed type

Fica confuso de ler como está agora

Author: aaaah sim, okok

```elixir
%{
  splits: [
    %{
      type: :fixed,
      amount: 12.5,

>       asset: "USD/2",

Reviewer: Não faria sentido repassar `USD/2` se o valor é `12.5`. Deveria ser só `USD`, ou `1250`.

      account: "some:destination:a"
    },
    %{
      type: :percent,
      amount: 30,
      remaining_to: "remaining:destination:b",
      splits: [
        %{
          type: :percent,
          amount: 40,
          account: "some:destination:c"
        },
        %{
          type: :percent,
          amount: 20,
          account: "some:destination:b"
        }
      ]
    }
  ],
  remaining_to: "remaining:destination:a",
  percent_asset: "USD/2"
}
```

> Will build:

Reviewer: ```suggestion
Will generate:
```

```
send [USD/2 *] (
  source = @user
  destination = {
    30% to {
      40% to @some:destination:c
      20% to @some:destination:b
      remaining to remaining:destination:b
    }
    remaining to remaining:destination:a
  }
)

send [USD/2 1250] (
  source = @user
  destination = @some:destination:a
)
```
Note that the `asset` field inside of nested destinations will be ignored, given that
they all go under the same `send` scope.

### Example 3

> Nested percent type plus fixed, and with fixed inside nests

Reviewer: Adicionar comas

      amount: 20,
      account: "some:destination:b"
    },
    %{
      type: :percent,
      amount: 30,
      remaining_to: "remaining:destination:b",
      splits: [
        %{
          type: :fixed,
          amount: 12.5,
          asset: "USD/2",
          account: "some:destination:c"
        },
        %{
          type: :percent,
          amount: 15,
          account: "some:destination:e"
        },
        %{
          type: :percent,
          amount: 20,
          account: "some:destination:d"
        },
        %{
          type: :percent,
          amount: 50,
          remaining_to: "remaining:destination:c",
          splits: [
            %{
              type: :fixed,
              amount: 5,
              asset: "BRL/2",
              account: "some:destination:b"
            },
            %{
              type: :percent,
              amount: 20,
              account: "some:destination:a"
            },
            %{
              type: :percent,
              amount: 27,
              account: "some:destination:a"
            }
          ]
        }
      ]
    }
  ],
  remaining_to: "remaining:destination:a",
  percent_asset: "EUR/2"
}
```

> Will build:

Reviewer: ```suggestion
Will generate:
```

```
send [EUR/2 *] (
  source = @user
  destination = {
    20% to @some:destination:b
    30% to {
      15% to @some:destination:e
      20% to @some:destination:d
      50% to {
        20% to @some:destination:a
        27% to @some:destination:a
        remaining to @remaining:destination:c
      }
      remaining to @remaining:destination:b
    }
    remaining to @remaining:destination:a
  }
)

send [USD/2 1250] (
  source = @user
  destination = @some:destination:c
)

send [BRL/2 500] (
  source = @user
  destination = @some:destination:b
)

send [EUR/2 1050] (
  source = @user
  destination = @some:destination:a
)
```
Note that the `fixed` types inside nests will be popped out and still be build after

> the `percent` types.

Reviewer: Não da para entender muito bem o que quer dizer aqui

# Builder

> This is a feature that makes possible to build numscripts dynamically within your application.

Reviewer: ```suggestion
This is a feature that makes it possible to build numscripts dynamically within your application.
```

%{
  splits: [
    %{
      type: :fixed,
      amount: 12.5,
      asset: "USD",
      account: "some:destination:a"
    },
    %{
      type: :percent,
      amount: 30,
      remaining_to: "remaining:destination:b",
      splits: [
        %{
          type: :percent,
          amount: 40,
          account: "some:destination:c"
        },
        %{
          type: :percent,
          amount: 20,
          account: "some:destination:b"
        }
      ]
    }
  ],
  remaining_to: "remaining:destination:a",
  percent_asset: "USD"
}
```
Will generate:
```
send [USD *] (
  source = @user
  destination = {
    30% to {
      40% to @some:destination:c
      20% to @some:destination:b
      remaining to remaining:destination:b
    }
    remaining to remaining:destination:a
  }
)

send [USD 1250] (
  source = @user
  destination = @some:destination:a
)
```
Note that the `asset` field inside of nested destinations will be ignored, given that
they all go under the same `send` scope.

### Example 3
Nested percent type plus fixed, and with fixed inside nests
```elixir
%{
  splits: [
    %{
      type: :fixed,
      amount: 10.5,
      asset: "EUR",
      account: "some:destination:a"

>     },

Reviewer: Fiquei pensando: e se for importante para essa moeda ter mais de 2 decimais? Como podemos especificar que usa 3? ou 4?

Author: acho que dá pro campo `asset` ser um map em vez de uma string, por ex:

```elixir
    %{
      type: :fixed,
      amount: 1050,
      asset: %{name: "EUR", decimal_places: 2}
      account: "some:destination:a"
    }
```

e depois concatenar os dois
```elixir
"#{asset.name}/#{asset.decimal_places}
```

e acho que vou fazer uma struct, ta ficando muito grande e meio confuso

Reviewer: Como é um valor externo (para ser usado com JSON) acho que não ajudaria muito ter um struct.

Seria bom se `decimal_places` for opcional, ou se usar só string (exemplo: "BRL") o padrão seja `2`, para ter a oportunidade de enviar o valor específico mas também não precisar definir sempre.

Author: Acho que só usar string facilita pra quem tá usando, e fazer um parse nessa string pra ver quantas casas decimais (ou não, se não tiver) não é nada muito complexo pra fazer do nosso lado.

E dois pontos importantes aqui:

1. Da pra colocar literalmente qualquer coisa nesse asset porque o Numscript não faz um parse pra ver se é um asset válido ou não
2. O comportamento do script é o mesmo independe das casas decimais e/ou do asset em si

Isso das casas decimais seria mais pra algo que a gente quisesse fazer internamente, como a feature de mostrar o valor decimal nas sctructs `Numscriptex.Posting` e `Numscriptex.Balance`

              account: "some:destination:b"
            },
            %{
              type: :percent,
              amount: 20,
              account: "some:destination:a"
            },
            %{
              type: :percent,
              amount: 27,
              account: "some:destination:a"
            }
          ]
        }
      ]
    }
  ],
  remaining_to: "remaining:destination:a",
  percent_asset: "EUR"
}
```
Will generate:
```
send [EUR *] (
  source = @user
  destination = {
    20% to @some:destination:b
    30% to {
      15% to @some:destination:e
      20% to @some:destination:d
      50% to {
        20% to @some:destination:a
        27% to @some:destination:a
        remaining to @remaining:destination:c
      }
      remaining to @remaining:destination:b
    }
    remaining to @remaining:destination:a
  }
)

send [USD 1250] (
  source = @user
  destination = @some:destination:c
)

send [BRL 500] (
  source = @user
  destination = @some:destination:b
)

send [EUR 1050] (
  source = @user
  destination = @some:destination:a
)
```
Note that the `fixed` types in nested `splits` (i.e. `splits` fields inside of `splits` fields) will be popped out and still

> be builded after the `percent` types.

Reviewer: ```suggestion
be built after the `percent` types.
```

You just need to call `Numscriptex.Builder.build/1` with the metadata necessary to build your numscript.

### Metadata
Metadata is the sole argument you need to use the `build/1` function, wich has the following fields:

Required fields:
- splits: a list of metadata needed to build numscripts (explained below)

Optional fields:
- percent_asset: monetary asset (BRL, USD, EUR, and etc.) to use when build numscripts using percentage values.
- remaining_to: where the rest of the money will go after the transaction is made
  - Only needed if you want to designate a specific destination. The default is [remaining kept](https://docs.formance.com/numscript/reference/destinations#allocation-destinations) wich makes the remainder go back to the source.

About the "list of metadata" mentioned on the `split` field above:

Required fields:
- account: account whose the money will go to
- amount: amount of money transferred (accepts both integer and float), but be aware that:
  - this field is treated as the literal amount you want to send (e.g. if yout want to send $5, your amount field must be 5 or 5.0)

>   - Numscript treats its amounts as integers insted of literals like `Numscriptex`(e.g. $5 would be 500), so the script generated will send 500 instead of 5.

Author: Depois de explicar isso aqui não sei se é lá uma boa UX, já que valor monetario no input é tratado de um jeito e no output de outro totalmente diferente. Acha melhor deixar como integer ($5 como 500) ou como ta mesmo?

Reviewer: Se for para ter sempre igual, acho melhor ter tudo como integer e esquecer dos decimais. É mais fácil usar decimal, mas complicaria muito.
